modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,337 @@
+  for (unsigned int i = 0; i < boundary_node_assm_size_0; i++)
+  {
+    // Find nodes to expand the chain
+    dof_id_type end_node_id = boundary_orderred_node_list.back();
+    auto isMatch1 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair) {
+      return old_id_pair.first == end_node_id;
+    };
+    auto isMatch2 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair) {
+      return old_id_pair.second == end_node_id;
+    };
+    auto result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch1);
+    bool match_first;
+    if (result == boundary_node_assm.end())
+    {
+      match_first = false;
+      result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch2);
+    }
+    else
+    {
+      match_first = true;
+    }
+    // If found, add the node to boundary_orderred_node_list
+    if (result != boundary_node_assm.end())
+    {
+      boundary_orderred_node_list.push_back(match_first ? (*result).second : (*result).first);
+      boundary_node_assm.erase(result);
+    }
+    // If there are still elements in boundary_node_assm and result ==
+    // boundary_node_assm.end(), this means the boundary is not a loop, the
+    // boundary_orderred_node_list is flipped and try the other direction that has not
+    // been examined yet.
+    else
+    {
+      if (isFlipped)
+      {
+        // Flipped twice; this means the boundary has at least two segments.
+        return false;
+      }
+      // mark the first flip event.
+      isFlipped = true;
+      std::reverse(boundary_orderred_node_list.begin(), boundary_orderred_node_list.end());
+      // As this iteration is wasted, set the iterator backward
+      i--;
+    }
+  }
+  // If the code ever gets here, boundary_node_assm is empty.
+  // If the isFlipped == true, the boundary is not a loop.
+  // This is not done inside the loop just for some potential applications in the future.
+  if (isFlipped)
+    return false;
+  // It the boundary is a loop, check if azimuthal angles change monotonically
+  else
+  {
+    // Utilize cross product here.
+    // If azimuthal angles change monotonically,
+    // the z components of the cross products are always negative or positive.
+    std::vector<Real> orderred_node_azi_list;
+    for (unsigned int i = 0; i < boundary_orderred_node_list.size() - 1; i++)
+    {
+      orderred_node_azi_list.push_back(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt)
+              .cross(*mesh.node_ptr(boundary_orderred_node_list[i + 1]) - origin_pt)(2));
+      // Use this opportunity to calculate maximum radius
+      max_node_radius = std::max(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt).norm(), max_node_radius);
+    }
+    std::sort(orderred_node_azi_list.begin(), orderred_node_azi_list.end());
+    if (orderred_node_azi_list.front() * orderred_node_azi_list.back() < 0.0)

```

Comment:
so in the case that two nodes and the origin are aligned, we get 0 there right? We cant really make the mesh then right? so maybe let's have a tolerance on that 0 ?

Reply:
If two nodes and the origin are aligned, it should be sure that there are positive AND negative elements in `orderred_node_azi_list` and this one would be < 0.0 (because the origin is the centroid, there should be nodes on both sides of that line.)